### PR TITLE
Issues/85 fhir reverse proxy server context path

### LINKED ISSUE #85: Make FHIR proxy server context path customizable
The context path of the FHIR proxy server is currently hardcoded to `/fhir`. To render this customizable, only a few changes need to be performed:

- Add an environment variable `APP_SERVER_CONTEXT_PATH` and default it to `/fhir` in `dsf-docker/fhir_proxy/Dockerfile`:
  ```diff
  ENV PROXY_PASS_CONNECTION_TIMEOUT_WS=30
  +ENV APP_SERVER_CONTEXT_PATH="/fhir"
  ```
- Replace the hardcoded context path `/fhir` in `dsf-docker/fhir_proxy/conf/extra/host-ssl.conf` with that environment variable:
  ```diff
  -<Location "/fhir">
  +<Location "${APP_SERVER_CONTEXT_PATH}">
          RequestHeader set X-ClientCert %{SSL_CLIENT_CERT}s
 
  -       ProxyPass http://${APP_SERVER_IP}:8080/fhir timeout=${PROXY_PASS_TIMEOUT_HTTP} connectiontimeout=${PROXY_PASS_CONNECTION_TIMEOUT_HTTP}
  -       ProxyPassReverse http://${APP_SERVER_IP}:8080/fhir
  +       ProxyPass http://${APP_SERVER_IP}:8080${APP_SERVER_CONTEXT_PATH} timeout=${PROXY_PASS_TIMEOUT_HTTP} connectiontimeout=${PROXY_PASS_CONNECTION_TIMEOUT_HTTP}
  +       ProxyPassReverse http://${APP_SERVER_IP}:8080${APP_SERVER_CONTEXT_PATH}
   </Location>
  -<Location "/fhir/ws">
  +<Location "${APP_SERVER_CONTEXT_PATH}/ws">
          RequestHeader set X-ClientCert %{SSL_CLIENT_CERT}s
 
          ProxyWebsocketFallbackToProxyHttp off
  -       ProxyPass ws://${APP_SERVER_IP}:8080/fhir/ws timeout=${PROXY_PASS_TIMEOUT_WS} connectiontimeout=${PROXY_PASS_CONNECTION_TIMEOUT_WS}
  -       ProxyPassReverse ws://${APP_SERVER_IP}:8080/fhir/ws
  +       ProxyPass ws://${APP_SERVER_IP}:8080${APP_SERVER_CONTEXT_PATH}/ws timeout=${PROXY_PASS_TIMEOUT_WS} connectiontimeout=${PROXY_PASS_CONNECTION_TIMEOUT_WS}
  +       ProxyPassReverse ws://${APP_SERVER_IP}:8080${APP_SERVER_CONTEXT_PATH}/ws
   </Location>
  ```
- Add the following (exemplarily) then to `docker-compose.yml`:
  (Note that the environment variable `DEV_DSF_SERVER_CONTEXT_PATH` already exists in `dsf-fhir/dsf-fhir-server-jetty/docker/Dockerfile`)
  ```diff
  services:
    proxy:
      image: ghcr.io/datasharingframework/fhir_proxy:1.1.0
      ...
      environment:
        ...
  +     APP_SERVER_CONTEXT_PATH: /dsf/fhir
    ...
    app:
      image: ghcr.io/datasharingframework/fhir:1.1.0
      ...
      environment:
        ...
  +     DEV_DSF_SERVER_CONTEXT_PATH: /dsf/fhir
    ...
  ```

The above has been implemented and successfully tested in our local environment already.

⚠️ There just seems to be a tiny bug though in `dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/adapter/HtmlFhirAdapter.java` in `private String getUrlHeading(Resource resource)` which leads to a a double display of `/fhir` in the shown title, e.g. `https://diz.uks.eu/dsf/fhir/fhir/metadata` and  - if I am not mistaken - there should be a `html` tab left to the `json` and `xml` tabs. (see screenshot)

temp

### PR DESCRIPTION
**Pull request by hhund**
* Adds reverse proxy config for requests to the context-url without trailing slash.
* Fixes url title in the UI for the "/" (root) and "/metadata" urls for context-paths other then `/fhir`.
* Improves some documentation.

closes #85

### CODE REVIEW

File: dsf-docker/fhir_proxy/README.md [hunk 1]
```diff
@@ -1,15 +1,48 @@
...
+
+**SSL_CA_CERTIFICATE_FILE**  
+Description: To set apache config param `SSLCACertificateFile`, with the trusted full CA chains for validating client certificates.  
+
+**SSL_CA_DN_REQUEST_FILE**  
+Description: To set apache config param `SSLCADNRequestFile` with client certificate signing CAs to modify the "Acceptable client certificate CA names" send to the client.  
+Default: All CAs from **SSL_CA_DN_REQUEST_FILE** if not set.  
+
+**SSL_VERIFY_CLIENT**  
+Description: To set apache config param `SSLVerifyClient `, default value `require`, set to `optional` when using OIDC authentication.
```
Comment (wetret) [inline dsf-docker/fhir_proxy/README.md:28]: Remove whitespace at the end of `SSLVerifyClient `

File: dsf-docker/fhir_proxy/conf/extra/host-ssl.conf [hunk 2]
```diff
@@ -38,8 +38,14 @@ Header always set Strict-Transport-Security "max-age=63072000; includeSubDomains
 <Location "${SERVER_CONTEXT_PATH}">
 	RequestHeader set X-ClientCert %{SSL_CLIENT_CERT}s
 
-	ProxyPass http://${APP_SERVER_IP}:8080/fhir timeout=${PROXY_PASS_TIMEOUT_HTTP} connectiontimeout=${PROXY_PASS_CONNECTION_TIMEOUT_HTTP}
-	ProxyPassReverse http://${APP_SERVER_IP}:8080/fhir
+	ProxyPass http://${APP_SERVER_IP}:8080/fhir/ timeout=${PROXY_PASS_TIMEOUT_HTTP} connectiontimeout=${PROXY_PASS_CONNECTION_TIMEOUT_HTTP}
```
Comment (wetret) [inline dsf-docker/fhir_proxy/conf/extra/host-ssl.conf:41]: Is it correct to hardcode `/fhir` as context path in `http://${APP_SERVER_IP}:8080/fhir/` or should this be the value of `SERVER_CONTEXT_PATH` resulting in `http://${APP_SERVER_IP}:8080/${SERVER_CONTEXT_PATH}/`?

Same question for the other following paths with hardcoded `/fhir` value.

**Review by wetret — APPROVED** [review]

File: dsf-docker/fhir_proxy/README.md [hunk 3]
```diff
@@ -1,15 +1,48 @@
...
+
+**SSL_CERTIFICATE_CHAIN_FILE**  
+Description: To set apache config param `SSLCertificateChainFile`, with the server certificate ca chain for (excluding the root CA), can be used if CA chain is not included in **SSL_CERTIFICATE_FILE**.  
+
+**SSL_CA_CERTIFICATE_FILE**  
+Description: To set apache config param `SSLCACertificateFile`, with the trusted full CA chains for validating client certificates.  
+
+**SSL_CA_DN_REQUEST_FILE**  
+Description: To set apache config param `SSLCADNRequestFile` with client certificate signing CAs to modify the "Acceptable client certificate CA names" send to the client.  
+Default: All CAs from **SSL_CA_DN_REQUEST_FILE** if not set.
```
Comment (schwzr) [inline dsf-docker/fhir_proxy/README.md:25]: If `SSL_CA_DN_REQUEST_FILE` is not set, the value of the same config option is used?

Comment (hhund) [reply dsf-docker/fhir_proxy/README.md]: Should say `SSL_CA_CERTIFICATE_FILE`. Fixed in 51370f1

File: dsf-docker/fhir_proxy/README.md (hunk 1 above)
Comment (hhund) [reply dsf-docker/fhir_proxy/README.md]: Fixed in 51370f1

**Pull request merged**